### PR TITLE
Reduce allocation in indexing an AbstractQ matrix

### PR DESCRIPTION
**Pull request by jishnub**
Currently the indexing operation `Q[i,j]` is computed as `ei ⋅ Q ⋅ ej`, where `ei` and `ej` are unit vectors. Both the unit vectors were allocated as `Vector`s. This PR changes the indexing to `(Q ⋅ ej)[i]`, which eliminates the allocation of `ei`. Since the result of `Q ⋅ ej` is overwritten into `ej`, the indexing `ej[i]` should be fast.

Some benchmarks:
```julia
julia> function getindex2(Q, i::Integer, j::Integer) # the updated method definition in this PR
           y = zeros(eltype(Q), size(Q, 2))
           y[j] = 1
           lmul!(Q, y)
           return y[i]
       end
getindex2 (generic function with 1 method)

julia> Q = qr(rand(30,30));

julia> @btime $Q.Q[1,1]
  2.044 μs (4 allocations: 944 bytes)
-0.07734152668340077

julia> @btime getindex2($Q.Q, 1, 1)
  1.958 μs (3 allocations: 640 bytes)
-0.07734152668340077
```

The performance improves marginally, and allocations go down.

### CODE REVIEW

**Comment by jishnub** [comment]
Test failure on macos64 seems unrelated (`ProcessExitedException`)

**Review by andreasnoack — APPROVED** [review]

**Comment by dkarrasch** [comment]
Shall we add a `getindex(Q, ::Colon, j::Int)` method? We could then define `getindex(Q, i, j) = getindex(Q, :, j)[i]`. Not sure about the "getrow" method.

**Comment by jishnub** [comment]
I have come up with a hacky "getrow" method that is unfortunately approximate at best:

```julia
function getindex(Q::AbstractQ, i::Int, ::Colon)
    x = zeros(eltype(Q), size(Q, 1))
    x[i] = 1
    lmul!(Q', x)
    vec(x')
end
```

Unfortunately this doesn't lead to `Q[i, :] == Matrix(Q)[i, :]`, so I have removed this for now.

**Comment by dkarrasch** [comment]
What about `rmul!(x', Q); return x`? Is that any better? Perhaps the issue is that we don't have a special `rmul!(::AbstractArray, Q)` method. After all, `x'` is not a strided array I guess.

**Comment by jishnub** [comment]
Why do tests fail? `qr` tests pass for me locally

File: stdlib/LinearAlgebra/src/qr.jl [hunk 1]
```diff
@@ -567,14 +567,18 @@ size(F::Union{QR,QRCompactWY,QRPivoted}) = size(getfield(F, :factors))
 size(Q::AbstractQ, dim::Integer) = size(getfield(Q, :factors), dim == 2 ? 1 : dim)
 size(Q::AbstractQ) = size(Q, 1), size(Q, 2)
 
-function getindex(Q::AbstractQ, i::Integer, j::Integer)
-    x = zeros(eltype(Q), size(Q, 1))
-    x[i] = 1
+copy(Q::AbstractQ) = I(size(Q, 1)) * Q
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/qr.jl:570]: What was wrong about `Matrix(Q)`? The fact that it doesn't necessarily produce a square matrix? Then I'd actually suggest
```suggestion
copy(Q::AbstractQ{T}) where {T} = lmul!(Q, Matrix{T}(I, size(Q)))
```
The fallback method of `copy(Q)` is superslow for larger `Q`. My suggestion follows our definition of `Matrix(Q)`, but makes sure that you get a square matrix, just as `copy(Q)` gives right now.

**Review by dkarrasch — APPROVED** [review]

**Comment by dkarrasch** [comment]
Failures look horrible, but unrelated. Let's run CI once more.

**Comment by DilumAluthge** [comment]
We've rebooted the Windows machines, so the Windows packagers should be working better now.

Regarding the `LinearAlgebra/tridiag` segfaults, the fix for that is now on master.

@jishnub Can you rebase your PR on the latest master?

**Pull request merged**